src/views/cdi-upload-provider/upload-pvc-form/UploadPVC.tsx

```diff
@@ -0,0 +1,275 @@
+import { match, useHistory } from 'react-router-dom';
+import axios from 'axios';
+import cx from 'classnames';
+
+import {
+  modelToGroupVersionKind,
+  PersistentVolumeClaimModel,
+  StorageClassModel,
+  TemplateModel,
+  V1Template,
+} from '@kubevirt-ui/kubevirt-api/console';
+import DataVolumeModel from '@kubevirt-ui/kubevirt-api/console/models/DataVolumeModel';
+import { V1beta1DataVolume } from '@kubevirt-ui/kubevirt-api/containerized-data-importer/models';
+import { IoK8sApiStorageV1StorageClass } from '@kubevirt-ui/kubevirt-api/kubernetes';
+import ExternalLink from '@kubevirt-utils/components/ExternalLink/ExternalLink';
+import { createUploadPVC } from '@kubevirt-utils/hooks/useCDIUpload/utils';
+import { useKubevirtTranslation } from '@kubevirt-utils/hooks/useKubevirtTranslation';
+import {
+  TEMPLATE_TYPE_BASE,
+  TEMPLATE_TYPE_LABEL,
+  TEMPLATE_VM_COMMON_NAMESPACE,
+} from '@kubevirt-utils/resources/template';
+import {
+  K8sVerb,
+  useAccessReview,
+  useK8sWatchResource,
+  WatchK8sResource,
+} from '@openshift-console/dynamic-plugin-sdk';
+import { ActionGroup, Alert, Button, ButtonVariant } from '@patternfly/react-core';
+
+import useBaseImages from '../hooks/useBaseImages';
+import useMultipleAccessReviews from '../hooks/useMultipleAccessReviews';
+import {
+  CDI_UPLOAD_OS_URL_PARAM,
+  CDI_UPLOAD_SUPPORTED_TYPES_URL,
+  CDI_UPLOAD_URL_BUILDER,
+  uploadErrorType,
+} from '../utils/consts';
+import { CDIUploadContext } from '../utils/context';
+import { getName, getNamespace, getPVCNamespace } from '../utils/selectors';
+import { resourcePath } from '../utils/utils';
+
+import UploadPVCButtonBar from './UploadPVCButtonBar';
+import UploadPVCForm from './UploadPVCForm';
+import UploadPVCFormStatus from './UploadPVCFormStatus';
+
+type UploadPVCPageProps = {
+  match: match<{ ns?: string }>;
+};
+
+const templatesResource: WatchK8sResource = {
+  isList: true,
+  optional: true,
+  groupVersionKind: modelToGroupVersionKind(TemplateModel),
+  namespace: TEMPLATE_VM_COMMON_NAMESPACE,
+  selector: {
+    matchLabels: { [TEMPLATE_TYPE_LABEL]: TEMPLATE_TYPE_BASE },
+  },
+};
+const UploadPVCPage: FC<UploadPVCPageProps> = (props) => {
+  const { t } = useKubevirtTranslation();
+  const [isSubmitting, setIsSubmitting] = useState(false);
+  const [isCheckingCertificate, setCheckingCertificate] = useState(false);
+  const [disableFormSubmit, setDisableFormSubmit] = useState(false);
+  const [fileValue, setFileValue] = useState<File>(null);
+  const [fileName, setFileName] = useState('');
+  const [fileNameExtension, setFileNameExtension] = useState('');
+  const [isFileRejected, setIsFileRejected] = useState(false);
+  const [error, setError] = useState<string>('');
+  const [isAllocating, setIsAllocating] = useState(false);
+  const [dvObj, setDvObj] = useState<V1beta1DataVolume>(null);
+  const [commonTemplates, loadedTemplates, errorTemplates] =
+    useK8sWatchResource<V1Template[]>(templatesResource);
```

Comment:
Sooo much states and we do also `setFileValue` `setFileName` and `setFileNameExtension` etc etc in combination. Using a `useReducer` will make the code cleaner. WDYT? In the near future maybe.